src/Storages/ObjectStorage/DataLakes/Iceberg/IcebergMetadata.cpp

```diff
@@ -390,34 +397,128 @@ IcebergSnapshot IcebergMetadata::getSnapshot(const String & manifest_list_file)
+            || manifest_content.getPartitionTransforms()[i] == PartitionTransform::Void)
+        {
+            continue;
+        }
+        Int32 source_id = manifest_file_entry.getContent().getPartitionSourceIds()[i];
+        NameAndTypePair name_and_type = schema_processor.getFieldCharacteristics(schema_version, source_id);
+        size_t column_size = manifest_file_entry.getContent().getPartitionColumns()[i]->size();
+        if (specific_info.ranges.empty())
+        {
+            specific_info.ranges.resize(column_size);
+        }
+        else
+        {
+            assert(specific_info.ranges.size() == column_size);
+        }
+        specific_info.partition_names_and_types.push_back(name_and_type);
+        for (size_t j = 0; j < column_size; ++j)
+        {
+            specific_info.ranges[j].push_back(getPartitionRange(
+                manifest_content.getPartitionTransforms()[i],
+                static_cast<UInt32>(j),
+                manifest_content.getPartitionColumns()[i],
+                name_and_type.type));
+        }
+    }
+    return specific_info;
+}
+
+std::vector<bool> IcebergMetadata::getPruningMask(const ManifestFileEntry & manifest_file_entry, const ActionsDAG * filter_dag) const
 {
-    if (!current_snapshot)
+    if (filter_dag == nullptr)
     {
         return {};
     }
+    SpecificSchemaPartitionInfo specific_info = getSpecificPartitionInfo(manifest_file_entry, current_schema_id);
+    std::vector<bool> pruning_mask(specific_info.ranges.size(), true);
+
+    if (!specific_info.partition_names_and_types.empty())
+    {
+        ExpressionActionsPtr partition_minmax_idx_expr = std::make_shared<ExpressionActions>(
+            ActionsDAG(specific_info.partition_names_and_types), ExpressionActionsSettings::fromContext(getContext()));
+        const KeyCondition partition_key_condition(
+            filter_dag, getContext(), specific_info.partition_names_and_types.getNames(), partition_minmax_idx_expr);
+        for (size_t j = 0; j < specific_info.ranges.size(); ++j)
+        {
+            if (!partition_key_condition.checkInHyperrectangle(specific_info.ranges[j], specific_info.partition_names_and_types.getTypes())
+                     .can_be_true)
+            {
+                ProfileEvents::increment(ProfileEvents::IcebergPartitionPrunnedFiles);
+                pruning_mask[j] = false;
+            }
+            else
+            {
+                pruning_mask[j] = true;
+            }
+        }
+    }
+    return pruning_mask;
+}
 
-    if (cached_files_for_current_snapshot.has_value())
+Strings IcebergMetadata::getDataFilesImpl(const ActionsDAG * filter_dag) const
+{
+    if (!current_snapshot)
     {
-        return cached_files_for_current_snapshot.value();
+        return {};
     }
 
-    Strings data_files;
-    for (const auto & manifest_entry : current_snapshot->getManifestList().getManifestFiles())
+    auto data_files_getter = [&]()
```

Comment:
Let's avoid using lambdas liek this. It makes the code harder to understand. Either move code inside lambda into a separate function, or rewrite the code so it won't need this lambda. For example:

```c++
if (!filter_dag && cached_files_for_current_snapshot.has_value())
    return cached_files_for_current_snapshot.value();

Strings data_files;
for (const auto & manifest_entry : current_snapshot->getManifestList().getManifestFiles())
{
   ...
}

if (!filter_dag)
    cached_unprunned_files_for_current_snapshot = data_files;
return data_files;
```